### PR TITLE
Support for Question Bank Feature in Advanced Quizzes.

### PR DESCRIPTION
**Pull request by actuallyakash**
## Description
Added support for the Question Bank feature.

## How has this been tested?
Manually.

## Types of changes
New feature

## Checklist:
- [x] This PR requires and contains at least one changelog file. 
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

**Comment by eri-trabiccolo** [comment]
Hey @actuallyakash 
I'd rather keep the whole logic into the add-on, and add here hooks we can leverage on the add-on code.
So instead of:
https://github.com/gocodebox/lifterlms/pull/2473/files#diff-017a7ce23e722494c709a870e2ce836adbfcc579a0249e5ee1cf4372332b0ee0R452
we'll have:
`		if ( apply_filters( 'lifterlms_randomize_quiz', $randomize, $quiz ) ) {`

Then in the add-on code:
```
add_filter(  'lifterlms_randomize_quiz', 'maybe_randomize_quiz' );
function maybe_randomize_quiz( $randomize, $quiz ) {
	return $randomize || lms_parse_bool( $quiz->get( 'question_bank' ) );
}
```
stuff like that.

You see what I mean?

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
see
https://github.com/gocodebox/lifterlms/pull/2473#issuecomment-1613118704

**Comment by actuallyakash** [comment]
Got it @eri-trabiccolo 

I wanted to ask you this, thanks for the clarification. Will transfer the functionality there.

File: includes/admin/views/builder/question.php [hunk 1]
```diff
@@ -171,6 +171,18 @@
 				<# } #>
 			</div>
 
+			<# if ( 'yes' === data.collection.parent.attributes.question_bank ) { #>
+			<div class="llms-settings-row">
+				<div class="llms-editable-toggle-group">
+					<label class="llms-switch">
+						<span class="llms-label"><?php _e( 'Include in Questions Bank', 'lifterlms' ); ?></span>
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/question.php]:
```suggestion
						<span class="llms-label"><?php esc_html_e( 'Include in Questions Bank', 'lifterlms' ); ?></span>
```

File: includes/models/model.llms.quiz.attempt.php [hunk 2]
```diff
@@ -437,17 +439,39 @@ private function get_new_questions() {
...
-					'earned'  => 0,
-					'points'  => $question->supports( 'points' ) ? $question->get( 'points' ) : 0,
-					'answer'  => null,
-					'correct' => null,
+					'id'                      => $question->get( 'id' ),
+					'earned'                  => 0,
+					'points'                  => $question->supports( 'points' ) ? $question->get( 'points' ) : 0,
+					'answer'                  => null,
+					'correct'                 => null,
+					'include_bank_variations' => 'yes' === $question->get( 'include_bank_variations' ) ? true : false,
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]: Hey Akash, what is this about? why is it here and not in the add-on?

File: includes/models/model.llms.quiz.php [hunk 3]
```diff
@@ -108,6 +109,27 @@ public function get_questions( $return = 'questions' ) {
...
+
+			$questions_limit = $this->get( 'number_of_questions' );
+
+			if ( $questions_limit ) {
+				return $questions_limit;
+			}
+		}
+
+		return count( $this->get_questions( 'ids' ) );
+	}
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.php]: Most if the logic should be moved into the add-on.
Here you can create a filter to be leveraged from the add-on.

File: templates/quiz/meta-information.php [hunk 4]
```diff
@@ -39,7 +40,7 @@
 	</li>
 
 	<li class="llms-quiz-meta-item llms-question-count">
-		<?php printf( __( 'Questions: %s', 'lifterlms' ), '<span class="llms-question-count">' . count( $quiz->get_questions( 'ids' ) ) . '</span>' ); ?>
+		<?php printf( __( 'Questions: %s', 'lifterlms' ), '<span class="llms-question-count">' . $quiz->get_questions_count() . '</span>' ); ?>
```
Comment (eri-trabiccolo) [inline templates/quiz/meta-information.php]: let's improve the escaping while we're here:
```suggestion
		<?php printf( esc_html__( 'Questions: %s', 'lifterlms' ), '<span class="llms-question-count">' . esc_html( $quiz->get_questions_count() ) . '</span>' ); ?>
```

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
Chan you start changing what requested here?

File: includes/admin/views/builder/question.php (hunk 1 above)
Comment (actuallyakash) [reply includes/admin/views/builder/question.php]: it's fixed

File: includes/models/model.llms.quiz.attempt.php (hunk 2 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.attempt.php]: My bad, I've moved this in the add-on and will try to move everything in the addon now.

This is a setting for making a question required to be included in the question bank.

File: includes/models/model.llms.quiz.php (hunk 3 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.php]: moved it. thanks

File: templates/quiz/meta-information.php (hunk 4 above)
Comment (actuallyakash) [reply templates/quiz/meta-information.php]: done

File: includes/models/model.llms.quiz.php [hunk 5]
```diff
@@ -30,6 +30,7 @@
  * @since 4.0.0 Remove deprecated methods.
  * @since 4.2.0 Added a parameter to the `is_orphan()` method to deeply check the quiz is not really attached to any lesson.
  * @since 5.0.0 Remove previously deprecated method `LLMS_Quiz::get_lessons()`.
+ * @since [verion] Add `get_questions_count()` function.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.php]:
```suggestion
 * @since [version] Add `get_questions_count()` function.
```

File: includes/models/model.llms.quiz.php [hunk 6]
```diff
@@ -108,6 +109,26 @@ public function get_questions( $return = 'questions' ) {
...
+	 */
+	public function get_questions_count() {
+
+		/**
+		 * Filter the count of questions in a quiz.
+		 *
+		 * @since [version]
+		 *
+		 * @param int       $questions_count Number of questions in a quiz.
+		 * @param LLMS_Quiz $this Current quiz object.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.php]:
```suggestion
		 * @param int       $questions_count Number of questions in a quiz.
		 * @param LLMS_Quiz $quiz            Current quiz object.
```

File: templates/quiz/meta-information.php [hunk 7]
```diff
@@ -7,7 +7,8 @@
  * @since 3.9.0
  * @since 4.0.0 Unknown.
  * @since 4.17.0 Return early if accessed without a logged in user or the quiz can't be loaded from the `$post` global.
- * @version 4.17.0
+ * @since [version] Added function for getting Quiz's questions count.
```
Comment (eri-trabiccolo) [inline templates/quiz/meta-information.php]: Did you add a function here? You use a new function instead of...

File: includes/models/model.llms.quiz.php (hunk 6 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.php]: fixed.

File: templates/quiz/meta-information.php (hunk 7 above)
Comment (actuallyakash) [reply templates/quiz/meta-information.php]: sorry, fixed now.

File: .changelogs/question-bank-1.yml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+significance: minor
+type: added
```
Comment (eri-trabiccolo) [inline .changelogs/question-bank-1.yml]:
```suggestion
type: dev
```

File: .changelogs/question-bank-2.yml [hunk 9]
```diff
@@ -0,0 +1,3 @@
+significance: minor
+type: added
```
Comment (eri-trabiccolo) [inline .changelogs/question-bank-2.yml]:
```suggestion
type: dev
```

File: .changelogs/question-bank-3.yml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+significance: minor
+type: added
```
Comment (eri-trabiccolo) [inline .changelogs/question-bank-3.yml]:
```suggestion
type: dev
```

File: .changelogs/question-bank-4.yml [hunk 11]
```diff
@@ -0,0 +1,3 @@
+significance: minor
+type: added
```
Comment (eri-trabiccolo) [inline .changelogs/question-bank-4.yml]:
```suggestion
type: dev
```

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

File: .changelogs/question-bank-4.yml (hunk 11 above)
Comment (actuallyakash) [reply .changelogs/question-bank-4.yml]: updated all! thanks 😬

File: assets/js/builder/Schemas/Quiz.js [hunk 12]
```diff
@@ -67,8 +67,21 @@ define( [], function() {
...
 						tip: LLMS.l10n.translate( 'Display questions in a random order for each attempt. Content questions are locked into their defined positions.' ),
 						type: 'switch',
+						condition: function() {
+							return 'yes' === this.get( 'question_bank' ) ? false : true;
+						}
 			},
-				],
+				], [
+					{
+						attribute: 'number_of_questions',
```
Comment (eri-trabiccolo) [inline assets/js/builder/Schemas/Quiz.js]: why the attribute here is `number_of_questions`?

File: .changelogs/question-bank-4.yml [hunk 13]
```diff
@@ -0,0 +1,3 @@
+significance: added
+type: dev
+entry: Added filter `llms_quiz_questions_count` for updating quiz's question count.
```
Comment (eri-trabiccolo) [inline .changelogs/question-bank-4.yml]:
```suggestion
entry: Added filter `llms_quiz_questions_count` to filter the quiz's question count.
```

File: assets/js/llms-builder.js [hunk 14]
```diff
@@ -3224,7 +3224,7 @@ define( 'Collections/Questions',[ 'Models/Question' ], function( model ) {
  * Quiz Schema
  *
  * @since    3.17.6
- * @version  3.24.0
+ * @version  [version]
```
Comment (eri-trabiccolo) [inline assets/js/llms-builder.js]: this whole file shouldn't be in the PR, it's a compiled one.

File: includes/admin/views/builder/question.php [hunk 15]
```diff
@@ -3,7 +3,7 @@
  * Builder question view
  *
  * @since   3.16.0
- * @version 3.27.0
+ * @version [version]
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/question.php:8]: I would like to see a changelog here, can you?

File: includes/admin/views/builder/settings-fields.php [hunk 16]
```diff
@@ -3,7 +3,7 @@
  * Model Field Settings Template
  *
  * @since   3.17.0
- * @version 3.24.0
+ * @version [version]
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/settings-fields.php:8]: changelog here too please <3

File: includes/models/model.llms.quiz.attempt.php [hunk 17]
```diff
@@ -28,6 +28,7 @@
  *              In the `l10n()` method, made sure the status key exists to avoid trying to access to array's undefined index.
  *              Added the public method `get_siblings()`.
  * @since 4.3.0 Added `$type` property declaration.
+ * @since [version] Added `llms_quiz_questions_randomize`, `llms_quiz_questions` and `llms_question_array` filters.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]:
```suggestion
```
no changelog needed here.

File: includes/models/model.llms.quiz.attempt.php [hunk 18]
```diff
@@ -436,18 +438,51 @@ private function get_new_questions() {
...
-					'points'  => $question->supports( 'points' ) ? $question->get( 'points' ) : 0,
-					'answer'  => null,
-					'correct' => null,
+				/**
+				 * Filter each question to add extra parameters for the questions array.
+				 *
+				 * @since [version]
+				 *
+				 * @param array         $question The question array to be used for the quiz.
+				 * @param LLMS_Question $question The question object.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]: same param name

File: includes/models/model.llms.quiz.attempt.php [hunk 19]
```diff
@@ -436,18 +438,51 @@ private function get_new_questions() {
...
-
+			/**
+			 * Filter questions for the quiz.
+			 *
+			 * Sets the questions to be used for the quiz.
+			 *
+			 * @since [version]
+			 *
+			 * @param array     $questions The questions to be used for the quiz.
+			 * @param LLMS_Quiz $llms_quiz LLMS_Quiz instance.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]: don't need the prefix in the param name, right? `$quiz` is enough.

File: includes/models/model.llms.quiz.attempt.php [hunk 20]
```diff
@@ -436,18 +438,51 @@ private function get_new_questions() {
...
+
+			/**
+			 * Filter randomize value for quiz questions.
+			 *
+			 * Sets the randomize to boolean value based on `Randomize` and `Question Bank`.
+			 *
+			 * @since [version]
+			 *
+			 * @param boolean   $randomize The randomize boolean value.
+			 * @param LLMS_Quiz $llms_quiz LLMS_Quiz instance.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]: don't need the prefix in the param name, right? $quiz is enough.

File: includes/models/model.llms.quiz.php [hunk 21]
```diff
@@ -30,6 +30,7 @@
  * @since 4.0.0 Remove deprecated methods.
  * @since 4.2.0 Added a parameter to the `is_orphan()` method to deeply check the quiz is not really attached to any lesson.
  * @since 5.0.0 Remove previously deprecated method `LLMS_Quiz::get_lessons()`.
+ * @since [version] Add `get_questions_count()` function.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.php]:
```suggestion
```
not needed.
I know the other changelog items are confusing you but we stopped logging changes in the class dockblock except for big changes that are not tracked elsewhere. E.g. when your *remove* something. But when you add something that's already tracked by the `@since` of the method/var/hook etc.

File: includes/models/model.llms.quiz.php [hunk 22]
```diff
@@ -108,6 +109,26 @@ public function get_questions( $return = 'questions' ) {
...
+	 */
+	public function get_questions_count() {
+
+		/**
+		 * Filter the count of questions in a quiz.
+		 *
+		 * @since [version]
+		 *
+		 * @param int       $questions_count Number of questions in a quiz.
+		 * @param LLMS_Quiz $this            Current quiz object.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.php]: In this case use `$quiz` instead of `$this` in the docblock.
This docblock is going to create an actual documentation, e.g. https://developer.lifterlms.com/reference/hooks/llms_student_query_default_args/

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

File: assets/js/builder/Schemas/Quiz.js (hunk 12 above)
Comment (actuallyakash) [reply assets/js/builder/Schemas/Quiz.js]: Left by mistake.

It's of no use in this case. `id: 'question-bank'`, can also be removed.

File: includes/admin/views/builder/question.php (hunk 15 above)
Comment (actuallyakash) [reply includes/admin/views/builder/question.php:8]: done

File: includes/admin/views/builder/settings-fields.php (hunk 16 above)
Comment (actuallyakash) [reply includes/admin/views/builder/settings-fields.php:8]: done 👌

File: includes/models/model.llms.quiz.attempt.php (hunk 18 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.attempt.php]: fixed.

File: includes/models/model.llms.quiz.attempt.php (hunk 19 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.attempt.php]: fixed.

File: includes/models/model.llms.quiz.attempt.php (hunk 20 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.attempt.php]: yup, fixed.

File: includes/models/model.llms.quiz.php (hunk 21 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.php]: got it. Wrote this in my notes 📓

File: includes/models/model.llms.quiz.php (hunk 22 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.php]: got it.

File: assets/js/llms-builder.js (hunk 14 above)
Comment (actuallyakash) [reply assets/js/llms-builder.js]: I've reverted the changes.

File: includes/admin/views/builder/question.php [hunk 23]
```diff
@@ -1,9 +1,11 @@
...
 /**
- * Builder question view
+ * Builder question view.
  *
- * @since   3.16.0
- * @version 3.27.0
+ * @since 3.16.0
+ * @since [version] Added checkbox for including question in Question Bank.
+ *
+ * @version [version]
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/question.php]:
```suggestion
 * @since 3.16.0
 * @since 3.27.0 Unknown.
 * @since [version] Added checkbox for including question in Question Bank.
 * @version [version]
```

File: includes/admin/views/builder/settings-fields.php [hunk 24]
```diff
@@ -1,9 +1,11 @@
...
 /**
- * Model Field Settings Template
+ * Model Field Settings Template.
  *
- * @since   3.17.0
- * @version 3.24.0
+ * @since 3.17.0
+ * @since [version] Added support for 'hidden', 'disabled' and 'switch-multi-input' field types.
+ *
+ * @version [version]
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/settings-fields.php]:
```suggestion
 * @since 3.17.0
 * @since 3.24.0 Unknown.
 * @since [version] Added support for 'hidden', 'disabled' and 'switch-multi-input' field types.
 * @version [version]
```

File: includes/models/model.llms.quiz.attempt.php [hunk 25]
```diff
@@ -436,18 +437,51 @@ private function get_new_questions() {
...
+
+			/**
+			 * Filter randomize value for quiz questions.
+			 *
+			 * Sets the randomize to boolean value based on `Randomize` and `Question Bank`.
+			 *
+			 * @since [version]
+			 *
+			 * @param boolean   $randomize The randomize boolean value.
+			 * @param LLMS_Quiz $quiz      LLMS_Quiz instance.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]:
```suggestion
			 * @param bool      $randomize The randomize boolean value.
			 * @param LLMS_Quiz $quiz      LLMS_Quiz instance.
```

File: assets/js/builder/Views/SettingsFields.js [hunk 26]
```diff
@@ -352,6 +353,10 @@ define( [], function() {
 					defaults.input_type  = 'url';
 				break;
 
+				case 'disabled':
+					defaults.url = 'https://lifterlms.com/product/advanced-quizzes/?utm_source=LifterLMS%20Plugin&utm_medium=Quiz%20Builder%20Button&utm_campaign=Advanced%20Question%20Upsell&utm_content=3.16.0&utm_term=Questions%20Bank';
```
Comment (eri-trabiccolo) [inline assets/js/builder/Views/SettingsFields.js]: we're creating a type here with a default which applies to only one case, it's too specific, it's not a default.

File: includes/admin/views/builder/question.php [hunk 27]
```diff
@@ -171,6 +173,18 @@
...
+				<div class="llms-editable-toggle-group">
+					<label class="llms-switch">
+						<span class="llms-label"><?php esc_html_e( 'Include in Questions Bank', 'lifterlms' ); ?></span>
+						<input type="checkbox" name="include_bank_variations"<# if ( 'yes' === data.get( 'include_bank_variations' ) ) { print( ' checked' ) } #>>
+						<div class="llms-switch-slider"></div>
+					</label>
+				</div>
+			</div>
+			<# } #>
+
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/question.php]: I would like to keep the logic in the advanced add-on as much as we can.

what if we have a `<?php do_action( 'llms_builder_question_after_clarifications' ); ?>` that we can leverage from the add-on?
Or even better, can't we just leverage the existing `llms_builder_question_after_features` action hook?

File: assets/js/builder/Views/SettingsFields.js [hunk 28]
```diff
@@ -352,6 +353,10 @@ define( [], function() {
 					defaults.input_type  = 'url';
 				break;
 
+				case 'disabled':
```
Comment (eri-trabiccolo) [inline assets/js/builder/Views/SettingsFields.js]: Also, instead of disabled, can't we have an "upsell" type here? makes more sense.
"disabled" is like a feature of an upsell.

File: assets/js/builder/Schemas/Quiz.js [hunk 29]
```diff
@@ -67,8 +69,20 @@ define( [], function() {
...
 			},
-				],
+				], [
+					{
+						id: 'question-bank',
+						label: LLMS.l10n.translate( 'Question Bank' ),
+						tip: LLMS.l10n.translate( 'A question bank helps prevent cheating by allowing instructors to create assessments with randomized questions pulled from a bank of questions. (Available in Advanced Quizzes addon)' ),
+						type: 'disabled',
+						text: 'Get LifterLMS Advanced Quizzes',
+						visibility: window.llms.hooks.applyFilters( 'llms_qb_upsell_visibility', 'visible' ),
```
Comment (eri-trabiccolo) [inline assets/js/builder/Schemas/Quiz.js]: Instead of playing with this filter (hence on the field visibility), can't we just leverage the filter `llms_define_quiz_schema`, and replace this field with the proper one from the add-on?
This will also avoid the addition of the `visibility` at all. Right?

File: includes/admin/views/builder/settings-fields.php [hunk 30]
```diff
@@ -90,29 +98,37 @@
...
-									<# } #>
-									<# if ( field.hasOwnProperty( 'min' ) ) { #> min="{{{ field.min }}}" <# } #>
-									<# if ( field.hasOwnProperty( 'max' ) ) { #> max="{{{ field.max }}}" <# } #>
-									name="{{{ field.attribute }}}"
-									<# if ( field.placeholder ) { #> placeholder="{{{ field.placeholder }}}" <# } #>
-									type="{{{ field.input_type }}}"
-									value="{{{ data.model.get( field.attribute ) }}}"
-								>
+								<# field.inputs.forEach( input => { #>
+									<div class="llms-input-wrapper{{{ input.col ? ' llms-col llms-col-' + input.col : '' }}}">
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/settings-fields.php]: do we really need this additional `col` attribute? can't we just handle all this with CSS only?

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

File: assets/js/builder/Schemas/Quiz.js (hunk 29 above)
Comment (actuallyakash) [reply assets/js/builder/Schemas/Quiz.js]: Nice suggestion 💯 , thanks. Fixed it.

File: assets/js/builder/Views/SettingsFields.js (hunk 28 above)
Comment (actuallyakash) [reply assets/js/builder/Views/SettingsFields.js]: yeah, I agree. Updated now.

File: assets/js/builder/Views/SettingsFields.js (hunk 26 above)
Comment (actuallyakash) [reply assets/js/builder/Views/SettingsFields.js]: Not useful, agreed.
Moved the URL into the upsell control.

File: includes/admin/views/builder/question.php (hunk 27 above)
Comment (actuallyakash) [reply includes/admin/views/builder/question.php]: Done. moved the code to advance quizzes.

File: includes/admin/views/builder/settings-fields.php (hunk 30 above)
Comment (actuallyakash) [reply includes/admin/views/builder/settings-fields.php]: oh yeah, gonna handle it by passing the class to `field.classes`.

File: includes/models/model.llms.quiz.attempt.php [hunk 31]
```diff
@@ -436,18 +437,51 @@ private function get_new_questions() {
...
+			 *
+			 * @param array     $questions The questions to be used for the quiz.
+			 * @param LLMS_Quiz $quiz      LLMS_Quiz instance.
+			 */
+			$questions = apply_filters( 'llms_quiz_questions', $questions, $quiz );
+
+			/**
+			 * Filter randomize value for quiz questions.
+			 *
+			 * Sets the randomize to boolean value based on `Randomize` and `Question Bank`.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]:
```suggestion
```

File: includes/models/model.llms.quiz.attempt.php [hunk 32]
```diff
@@ -413,6 +413,7 @@ public function get_key() {
 	 *
 	 * @since 3.9.0
 	 * @since 3.16.0 Unknown.
+	 * @since [version] Added `llms_quiz_questions_randomize` and `llms_quiz_questions` filters and `include_bank_variations` parameter to questions array.
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]:
```suggestion
```
We don't need a changelog here.
Why?
> Added `llms_quiz_questions_randomize` and `llms_quiz_questions` filters

this is already covered by the `@since` tag for each filter.

> and `include_bank_variations` parameter to questions array.

There's nothing here about that, probably in the add-on?

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]
Looks good, minor changes requested.

Can you also add a changelog item for the addition of the method: `LLMS_Quiz::get_questions_count()`?
Additionally all the filters that you added in the quiz attempt model, should be better start with:
`llms_quiz_attempt_` rather than `llms_quiz_` only.
Even `llms_question_array` => `llms_quiz_attempt_question_array`.
Gives more context.
Also, let's always pass the attempt and the quiz objects as params.

File: includes/models/model.llms.quiz.attempt.php (hunk 32 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.attempt.php]: oh! got it. Will remember next time.

File: includes/models/model.llms.quiz.php [hunk 33]
```diff
@@ -108,6 +108,26 @@ public function get_questions( $return = 'questions' ) {
...
+
+		/**
+		 * Filter the count of questions in a quiz.
+		 *
+		 * @since [version]
+		 *
+		 * @param int       $questions_count Number of questions in a quiz.
+		 * @param LLMS_Quiz $quiz            Current quiz object.
+		 */
+		return apply_filters( 'llms_quiz_attempt_questions_count', count( $this->get_questions( 'ids' ) ), $this );
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.php]:
```suggestion
		return apply_filters( 'llms_quiz_questions_count', count( $this->get_questions( 'ids' ) ), $this );
```
there is no attempt here... of course

File: .changelogs/question-bank-4.yml [hunk 34]
```diff
@@ -0,0 +1,3 @@
+significance: added
+type: dev
+entry: Added filter `llms_quiz_attempt_questions_count` to filter the quiz's question count.
```
Comment (eri-trabiccolo) [inline .changelogs/question-bank-4.yml]: there is no attempt here... of course

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

File: includes/models/model.llms.quiz.php (hunk 33 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.php]: fixed!

File: .changelogs/question-bank-4.yml (hunk 34 above)
Comment (actuallyakash) [reply .changelogs/question-bank-4.yml]: fixed!

File: includes/models/model.llms.quiz.attempt.php [hunk 35]
```diff
@@ -436,18 +445,46 @@ private function get_new_questions() {
...
+			 *
+			 * Sets the questions to be used for the quiz.
+			 *
+			 * @since [version]
+			 *
+			 * @param array             $questions The questions to be used for the quiz.
+			 * @param LLMS_Quiz         $quiz      LLMS_Quiz instance.
+			 * @param LLMS_Quiz_Attempt $attempt   LLMS_Quiz_Attempt instance.
+			 */
+			$questions = apply_filters( 'llms_quiz_attempt_questions', $questions, $quiz, $this );
```
Comment (eri-trabiccolo) [inline includes/models/model.llms.quiz.attempt.php]: Why are you filtering the questions here and not some lines above?
`		foreach ( $quiz->get_questions() as $index => $question ) {`

What's the reasoning?

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

File: includes/models/model.llms.quiz.attempt.php (hunk 35 above)
Comment (actuallyakash) [reply includes/models/model.llms.quiz.attempt.php]: just to separate the logic of the filter from the above logic.
And after the foreach, I'm only getting the required details.

Comment (actuallyakash) [reply includes/models/model.llms.quiz.attempt.php]: hey @eri-trabiccolo 

I've updated the filter and moved it up before the loop.

File: assets/js/builder/Views/SettingsFields.js [hunk 36]
```diff
@@ -2,7 +2,7 @@
  * Model settings fields view
  *
  * @since 3.17.0
- * @version 4.7.0
+ * @version [version]
```
Comment (eri-trabiccolo) [inline assets/js/builder/Views/SettingsFields.js]:
```suggestion
 * @version 4.7.0
```
Nothing was changed here :D

File: includes/admin/views/builder/question.php (hunk 15 above)
Comment (eri-trabiccolo) [reply includes/admin/views/builder/question.php:8]: This was not done I don't see any changelog here.
What changed did you make to this file, can you summarize them?

File: includes/admin/views/builder/settings-fields.php [hunk 37]
```diff
@@ -1,9 +1,11 @@
 <?php
 /**
- * Model Field Settings Template
+ * Model Field Settings Template.
  *
- * @since   3.17.0
- * @version 3.24.0
+ * @since 3.17.0
+ * @since 3.24.0 Unknown.
+ * @since [version] Added support for Question Banks multiple input fields.
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/settings-fields.php]: and the upsell field type?

File: templates/quiz/meta-information.php [hunk 38]
```diff
@@ -7,7 +7,8 @@
  * @since 3.9.0
  * @since 4.0.0 Unknown.
  * @since 4.17.0 Return early if accessed without a logged in user or the quiz can't be loaded from the `$post` global.
- * @version 4.17.0
+ * @since [version] Used function `get_questions_count()` for showing count.
```
Comment (eri-trabiccolo) [inline templates/quiz/meta-information.php]: and output escaping

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

File: templates/quiz/meta-information.php (hunk 38 above)
Comment (actuallyakash) [reply templates/quiz/meta-information.php]: done.

File: assets/js/builder/Views/SettingsFields.js (hunk 36 above)
Comment (actuallyakash) [reply assets/js/builder/Views/SettingsFields.js]: yup (changed but reverted) 👨‍💻

File: includes/admin/views/builder/question.php (hunk 15 above)
Comment (actuallyakash) [reply includes/admin/views/builder/question.php:8]: I think I've committed the suggested changes and got confused about this. Now fixed.

File: includes/admin/views/builder/settings-fields.php (hunk 37 above)
Comment (actuallyakash) [reply includes/admin/views/builder/settings-fields.php]: done.

File: includes/admin/views/builder/question.php (hunk 15 above)
Comment (eri-trabiccolo) [reply includes/admin/views/builder/question.php:8]: Still to be fixed

File: includes/admin/views/builder/question.php [hunk 39]
```diff
@@ -1,9 +1,12 @@
...
- * Builder question view
+ * Builder question view.
  *
- * @since   3.16.0
- * @version 3.27.0
+ * @since 3.16.0
+ * @since 3.27.0 Unknown.
+ * @since [version] Escaped strings in labels.
+ *
+ * @version [version]
```
Comment (eri-trabiccolo) [inline includes/admin/views/builder/question.php]:
```suggestion
 * @since [version] Escaped strings in labels.
 * @version [version]
```

File: assets/js/builder/Models/Quiz.js [hunk 40]
```diff
@@ -46,14 +48,16 @@ define( [
...
-		 * New lesson defaults
-		 * @return   obj
-		 * @since    3.16.0
-		 * @version  3.16.6
+		 * New lesson defaults.
+		 *
+		 * @since 3.16.0
+		 * @since [version] Added filter for filtering defaults.
+		 *
+		 * @version [version]
```
Comment (eri-trabiccolo) [inline assets/js/builder/Models/Quiz.js]:
```suggestion
		 *
		 * @return {Object}
```
as you know we don't add the version tag anymore...

File: includes/admin/views/builder/settings-fields.php (hunk 37 above)
Comment (eri-trabiccolo) [inline includes/admin/views/builder/settings-fields.php]:
```suggestion
 * @since [version] Added support for `upsell` field type and multiple input fields.
```

File: templates/quiz/meta-information.php [hunk 41]
```diff
@@ -7,7 +7,8 @@
  * @since 3.9.0
  * @since 4.0.0 Unknown.
  * @since 4.17.0 Return early if accessed without a logged in user or the quiz can't be loaded from the `$post` global.
- * @version 4.17.0
+ * @since [version] Used function `get_questions_count()` for showing count and escaped labels.
```
Comment (eri-trabiccolo) [inline templates/quiz/meta-information.php]:
```suggestion
 * @since [version] Used `LLMS_Quiz::get_questions_count()` method for showing count and escaped labels.
```

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

**Review by eri-trabiccolo — APPROVED** [review]
Looks good, thanks.
Won't merge until the advanced quizzes add-on is ready for this, otherwise we'd have the upsell pointing to nothing :D

**Comment by actuallyakash** [comment]
> Looks good, thanks. Won't merge until the advanced quizzes add-on is ready for this, otherwise we'd have the upsell pointing to nothing :D

Got it :D 

Gonna wrap up advanced quizzes soon 👌

File: assets/js/builder/Models/Quiz.js [hunk 42]
```diff
@@ -1,7 +1,9 @@
 /**
- * Quiz Model
- * @since    3.16.0
- * @version  3.24.0
+ * Quiz Model.
+ *
+ * @since 3.16.0
+ *
+ * @version [version]
```
Comment (eri-trabiccolo) [inline assets/js/builder/Models/Quiz.js]:
```suggestion
 * @since 3.16.0
 * @version [version]
```

File: assets/js/builder/Schemas/Quiz.js [hunk 43]
```diff
@@ -1,8 +1,10 @@
 /**
- * Quiz Schema
+ * Quiz Schema.
  *
- * @since    3.17.6
- * @version  3.24.0
+ * @since 3.17.6
+ * @since [version] Added upsell for Question Bank.
+ *
+ * @version [version]
```
Comment (eri-trabiccolo) [inline assets/js/builder/Schemas/Quiz.js]:
```suggestion
 * @since 3.17.6
 * @since [version] Added upsell for Question Bank.
 * @version [version]
```
and you also made `random_question` conditionally visible.

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

File: .changelogs/question-bank-5.yml [hunk 44]
```diff
@@ -0,0 +1,4 @@
+significance: minor
+type: added
+entry: Added function `LLMS_Quiz::get_questions_count()` for getting count of
```
Comment (eri-trabiccolo) [inline .changelogs/question-bank-5.yml]:
```suggestion
entry: Added method `LLMS_Quiz::get_questions_count()` for getting count of
```

File: .changelogs/question-bank.yml [hunk 45]
```diff
@@ -0,0 +1,3 @@
+significance: minor
+type: added
+entry: Added support for Question Bank feature in LifterLMS Advanced Quizzes.
```
Comment (eri-trabiccolo) [inline .changelogs/question-bank.yml]:
```suggestion
entry: Added support for "Question Bank" feature of the LifterLMS Advanced Quizzes add-on.
```

**Review by eri-trabiccolo — CHANGES_REQUESTED** [review]

**Review by eri-trabiccolo — APPROVED** [review]
